environment_TO.py

```diff
@@ -0,0 +1,767 @@
+        '''
+        self.name = name
+        self.conf = conf
+        
+        self.nx = self.conf.nx
+        self.nu = self.conf.na
+
+        # Rename reward parameters
+        self.offset = self.conf.cost_funct_param[0]
+        self.scale = self.conf.cost_funct_param[1]
+
+        self.alpha = self.conf.soft_max_param[0]
+        self.alpha2 = self.conf.soft_max_param[1]
+
+        self.XC1 = self.conf.obs_param[0]
+        self.YC1 = self.conf.obs_param[1]
+        self.XC2 = self.conf.obs_param[2]
+        self.YC2 = self.conf.obs_param[3]
+        self.XC3 = self.conf.obs_param[4]
+        self.YC3 = self.conf.obs_param[5]
+        self.A1 = self.conf.obs_param[6]
+        self.B1 = self.conf.obs_param[7]
+        self.A2 = self.conf.obs_param[8]
+        self.B2 = self.conf.obs_param[9]
+        self.A3 = self.conf.obs_param[10]
+        self.B3 = self.conf.obs_param[11]
+
+        self.x_des = self.conf.TARGET_STATE[0]
+        self.y_des = self.conf.TARGET_STATE[1]
+
+        # The self.xdot will be a casadi function mapping:  state,control -> [velocity,acceleration]
+        cx = casadi.SX.sym("x",self.nx,1)
+        cu = casadi.SX.sym("u",self.nu,1)
+
+        self.x_next = casadi.Function('x_next', [cx, cu], [self.simulate_fun(cx,cu)])
+
+        self.p_ee = casadi.Function('p_ee', [cx], [self.get_end_effector_position_fun(cx)])
+
+        if self.name == 'running_model':
+            self.weights = np.copy(self.conf.cost_weights_running)
+        elif self.name == 'terminal_model':
+            self.weights = np.copy(self.conf.cost_weights_terminal)
+        else:
+            print("The model can be either 'running_model' or 'terminal_model'")
+            import sys
+            sys.exit()
+
+        self.cost = casadi.Function('cost', [cx,cu], [self.cost_fun(cx,cu)])
+
+        #self.step = casadi.Function('step', [cx,cu], [self.step_fun(cx,cu)])
+
+    def get_end_effector_position_fun(self, cx):        
+        return casadi.vertcat(cx[0],cx[1],0)
+    
+    def bound_control_cost(self, action):
+        u_cost = 0
+        for i in range(self.conf.nb_action):
+            u_cost += action[i]*action[i] + 1e2*(np.exp(-self.conf.w_b*(action[i]-self.conf.u_min[i])) + np.exp(-self.conf.w_b*(self.conf.u_max[i]-action[i])))
+
+        return u_cost
+    
+    def cost_fun(self, x, u):
+        ''' Compute cost '''
+        p_ee = self.p_ee(x)
+
+        ### Penalties representing the obstacle ###
+        ell1_cost = (np.log(np.exp(self.alpha*-(((p_ee[0]-self.XC1)**2)/((self.A1/2)**2) + ((p_ee[1]-self.YC1)**2)/((self.B1/2)**2) - 1.0)) + 1)/self.alpha)  
+        ell2_cost = (np.log(np.exp(self.alpha*-(((p_ee[0]-self.XC2)**2)/((self.A2/2)**2) + ((p_ee[1]-self.YC2)**2)/((self.B2/2)**2) - 1.0)) + 1)/self.alpha) 
+        ell3_cost = (np.log(np.exp(self.alpha*-(((p_ee[0]-self.XC3)**2)/((self.A3/2)**2) + ((p_ee[1]-self.YC3)**2)/((self.B3/2)**2) - 1.0)) + 1)/self.alpha)
```

Comment:
Isn't it possible to avoid having the same functions defined twice? (Once here and once in Environment.py)